src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs

```diff
@@ -1192,137 +1243,180 @@ static void ProcessNode(RegexNode node, RegexNode subsequent, int maxDepth)
-                                    case Notonelazy when subsequent.M > 0 && node.Ch == subsequent.Ch:
-                                    case Notoneloop when subsequent.M > 0 && node.Ch == subsequent.Ch:
-                                    case Notoneloopatomic when subsequent.M > 0 && node.Ch == subsequent.Ch:
-                                    case Multi when node.Ch != subsequent.Str![0]:
-                                    case Set when !RegexCharClass.CharInClass(node.Ch, subsequent.Str!):
-                                    case Setlazy when subsequent.M > 0 && !RegexCharClass.CharInClass(node.Ch, subsequent.Str!):
-                                    case Setloop when subsequent.M > 0 && !RegexCharClass.CharInClass(node.Ch, subsequent.Str!):
-                                    case Setloopatomic when subsequent.M > 0 && !RegexCharClass.CharInClass(node.Ch, subsequent.Str!):
-                                    case End:
-                                    case EndZ when node.Ch != '\n':
-                                    case Eol when node.Ch != '\n':
-                                    case Boundary when RegexCharClass.IsWordChar(node.Ch):
-                                    case Nonboundary when !RegexCharClass.IsWordChar(node.Ch):
-                                    case ECMABoundary when RegexCharClass.IsECMAWordChar(node.Ch):
-                                    case NonECMABoundary when !RegexCharClass.IsECMAWordChar(node.Ch):
-                                        return true;
-                                }
-                                break;
+        /// <summary>
+        /// Determines whether node can be switched to an atomic loop.  Subsequent is the node
+        /// immediately after 'node'.
+        /// </summary>
+        private static bool CanBeMadeAtomic(RegexNode node, RegexNode subsequent, int maxDepth)
+        {
+            if (maxDepth <= 0)
+            {
+                // We hit our recursion limit.  Just don't apply the optimization.
+                return false;
+            }
 
-                            case Notoneloop:
-                                switch (subsequent.Type)
-                                {
-                                    case One when node.Ch == subsequent.Ch:
-                                    case Onelazy when subsequent.M > 0 && node.Ch == subsequent.Ch:
-                                    case Oneloop when subsequent.M > 0 && node.Ch == subsequent.Ch:
-                                    case Oneloopatomic when subsequent.M > 0 && node.Ch == subsequent.Ch:
-                                    case Multi when node.Ch == subsequent.Str![0]:
-                                    case End:
-                                        return true;
-                                }
-                                break;
+            // Skip the successor down to the guaranteed next node.
```

Comment:
Not changed line, but it's not clear what is meant by "guaranteed next node". Maybe "closest node that's guaranteed to follow it"  ?